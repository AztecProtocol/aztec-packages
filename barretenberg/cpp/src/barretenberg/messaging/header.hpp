#pragma once
#include "barretenberg/serialize/msgpack.hpp"
#include <cstdint>
#include <cstring>

namespace bb::messaging {

enum SystemMsgTypes { TERMINATE = 0, PING = 1, PONG = 2 };

const uint32_t FIRST_APP_MSG_TYPE = 100;

// #pragma pack(push, 1)
struct MsgHeader {
    uint32_t messageId; // Unique Id for the message
    uint32_t requestId; // Id of the message this is responding too (may not be used)

    MSGPACK_FIELDS(messageId, requestId);

    MsgHeader() = default;

    MsgHeader(uint32_t reqId)
        : requestId(reqId)
    {}

    MsgHeader(uint32_t msgId, uint32_t reqId)
        : messageId(msgId)
        , requestId(reqId)
    {}
};

struct HeaderOnlyMessage {
    uint32_t msgType;
    MsgHeader header;

    HeaderOnlyMessage(uint32_t type, MsgHeader& hdr)
        : msgType(type)
        , header(hdr)
    {}

    HeaderOnlyMessage() = default;

    MSGPACK_FIELDS(msgType, header);
};

template <class T> struct TypedMessage {
    uint32_t msgType;
    MsgHeader header;
    T value;

    TypedMessage(uint32_t type, MsgHeader& hdr, const T& val)
        : msgType(type)
        , header(hdr)
        , value(val)
    {}

    TypedMessage() = default;

    MSGPACK_FIELDS(msgType, header, value);
};

// #pragma pack(pop)
} // namespace bb::messaging
