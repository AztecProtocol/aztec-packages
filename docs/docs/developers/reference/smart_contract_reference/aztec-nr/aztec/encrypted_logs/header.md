# EncryptedLogHeader

## Fields
| Field | Type |
| --- | --- |
| address | AztecAddress |

## Methods

### new

```rust
EncryptedLogHeader::new(address);
```

#### Parameters
| Name | Type |
| --- | --- |
| address | AztecAddress |

### compute_ciphertext

```rust
EncryptedLogHeader::compute_ciphertext(self, secret, pk);
```

#### Parameters
| Name | Type |
| --- | --- |
| self |  |
| secret | Scalar |
| pk | T |

## Standalone Functions

### test_encrypted_log_header_matches_noir

```rust
test_encrypted_log_header_matches_noir();
```

Takes no parameters.

