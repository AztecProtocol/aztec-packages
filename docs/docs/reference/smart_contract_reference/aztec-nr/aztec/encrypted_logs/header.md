# EncryptedLogHeader

## Fields
| Field | Type |
| --- | --- |
| address | AztecAddress |

## Methods

### new

```rust
EncryptedLogHeader::new(address);
```

#### Parameters
| Name | Type |
| --- | --- |
| address | AztecAddress |

### compute_ciphertext

@todo Issue(#5901) Figure out if we return the bytes or fields for the log

```rust
EncryptedLogHeader::compute_ciphertext(self, secret, point);
```

#### Parameters
| Name | Type |
| --- | --- |
| self |  |
| secret | GrumpkinPrivateKey |
| point | GrumpkinPoint |

## Standalone Functions

### test_encrypted_log_header

```rust
test_encrypted_log_header();
```

Takes no parameters.

