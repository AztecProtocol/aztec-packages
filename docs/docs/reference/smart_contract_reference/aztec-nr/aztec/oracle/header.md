## Standalone Functions

### get_header_at_oracle

```rust
get_header_at_oracle(_block_number);
```

#### Parameters
| Name | Type |
| --- | --- |
| _block_number | u32 |

### get_header_at_internal

```rust
get_header_at_internal(block_number);
```

#### Parameters
| Name | Type |
| --- | --- |
| block_number | u32 |

### get_header_at

```rust
get_header_at(block_number, context);
```

#### Parameters
| Name | Type |
| --- | --- |
| block_number | u32 |
| context | PrivateContext |

