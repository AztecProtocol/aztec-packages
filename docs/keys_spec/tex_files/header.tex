\usepackage[utf8]{inputenc}

\usepackage[T1]{fontenc}

\usepackage[english]{babel}

\usepackage{amsmath,amssymb,amsfonts}

\usepackage{mathrsfs,mathtools}

\usepackage[all]{xy} % for diagrams

\usepackage{bbm}

\usepackage{BOONDOX-calo} % for calligraphic font

%\usepackage{stmaryrd} % for llbracket

% \usepackage{algorithm,algpseudocode}
\usepackage{algorithm, algpseudocodex}

\usepackage{cite}

\usepackage{enumitem}

%\usepackage{tabularx}

%\usepackage{multirow}

\usepackage{xspace}

\usepackage{xcolor}

\usepackage{url}

\usepackage[colorlinks=false,urlcolor=black]{hyperref}

% For embedding images
\usepackage{graphicx}

\setcounter{secnumdepth}{3}

%%% ENVIRONMENTS

\spnewtheorem{thm}[definition]{Theorem}{\bfseries}{\itshape}
\spnewtheorem{lem}[thm]{Lemma}{\bfseries}{\itshape}
\spnewtheorem{cor}[thm]{Corollary}{\bfseries}{\itshape}
