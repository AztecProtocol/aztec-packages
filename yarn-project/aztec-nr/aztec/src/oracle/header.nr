use dep::std::merkle::compute_merkle_root;
use dep::protocol_types::{
    constants::HEADER_LENGTH,
    header::Header,
};

use crate::{
    context::PrivateContext,
    oracle::get_membership_witness::get_archive_membership_witness,
};

#[oracle(getHeader)]
fn get_header_at_oracle(_block_number: u32) -> [Field; HEADER_LENGTH] {}

unconstrained pub fn get_header_at_internal(block_number: u32) -> Header {
    let header = get_header_at_oracle(block_number);
    Header::deserialize(header)
}

pub fn get_header_at(block_number: u32, context: PrivateContext) -> Header {
    let historical_header_block_number = context.historical_header.global_variables.block_number as u32;

    if (block_number == historical_header_block_number) {
        // If the block number we want to prove against is the same as the block number in the historical header we
        // skip the inclusion proofs and just return the historical header from context.
        context.historical_header
    } else {
        // 1) Get block number corresponding to the last_archive root in the header
        // Note: We subtract 1 because the last_archive root is the root of the archive after applying the previous block
        let last_archive_block_number = historical_header_block_number - 1;

        // 2) Check that the last archive block number is more than or equal to the block number we want to prove against
        // We could not perform the proof otherwise because the last archive root from the header would not "contain"
        // the header we want to prove against
        assert(
            last_archive_block_number >= block_number, "Last archive block number is smaller than the block number we want to prove against"
        );

        // 3) Get the header of a given block from oracle
        let header = get_header_at_internal(block_number);

        // 4) Compute the block hash from the block header
        let block_hash = header.hash();

        // 5) Get the membership witness of the block in the archive
        let witness = get_archive_membership_witness(last_archive_block_number, block_hash);

        // 6) Check that the block is in the archive (i.e. the witness is valid)
        assert(
            context.historical_header.last_archive.root
            == compute_merkle_root(block_hash, witness.index, witness.path), "Proving membership of a block in archive failed"
        );

        // 7) Return the block header
        header
    }
}
