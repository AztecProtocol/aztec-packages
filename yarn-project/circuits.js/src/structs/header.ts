import { pedersenHash } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';

import { GeneratorIndex, HEADER_LENGTH } from '../constants.gen.js';
import { ContentCommitment } from './content_commitment.js';
import { GlobalVariables } from './global_variables.js';
import { AppendOnlyTreeSnapshot } from './rollup/append_only_tree_snapshot.js';
import { StateReference } from './state_reference.js';

/** A header of an L2 block. */
export class Header {
  constructor(
    /** Snapshot of archive before the block is applied. */
    public lastArchive: AppendOnlyTreeSnapshot,
    /** Hash of the body of an L2 block. */
    public contentCommitment: ContentCommitment,
    /** State reference. */
    public state: StateReference,
    /** Global variables of an L2 block. */
    public globalVariables: GlobalVariables,
  ) {}

  toBuffer() {
    // Note: The order here must match the order in the HeaderLib solidity library.
    return serializeToBuffer(this.lastArchive, this.contentCommitment, this.state, this.globalVariables);
  }

  toFields(): Fr[] {
    // Note: The order here must match the order in header.nr
    const fields = [
      ...this.lastArchive.toFields(),
      ...this.contentCommitment.toFields(),
      ...this.state.toFields(),
      ...this.globalVariables.toFields(),
    ];
    if (fields.length !== HEADER_LENGTH) {
      throw new Error(`Invalid number of fields for Header. Expected ${HEADER_LENGTH}, got ${fields.length}`);
    }
    return fields;
  }

  static fromBuffer(buffer: Buffer | BufferReader): Header {
    const reader = BufferReader.asReader(buffer);

    return new Header(
      reader.readObject(AppendOnlyTreeSnapshot),
      reader.readObject(ContentCommitment),
      reader.readObject(StateReference),
      reader.readObject(GlobalVariables),
    );
  }

  static fromFields(fields: Fr[] | FieldReader): Header {
    const reader = FieldReader.asReader(fields);

    const lastArchive = new AppendOnlyTreeSnapshot(reader.readField(), Number(reader.readField().toBigInt()));
    const contentCommitment = ContentCommitment.fromFields(reader);
    const state = StateReference.fromFields(reader);
    const globalVariables = GlobalVariables.fromFields(reader);

    return new Header(lastArchive, contentCommitment, state, globalVariables);
  }

  static empty(): Header {
    return new Header(
      AppendOnlyTreeSnapshot.zero(),
      ContentCommitment.empty(),
      StateReference.empty(),
      GlobalVariables.empty(),
    );
  }

  isEmpty(): boolean {
    return (
      this.lastArchive.isZero() &&
      this.contentCommitment.isEmpty() &&
      this.state.isEmpty() &&
      this.globalVariables.isEmpty()
    );
  }

  /**
   * Serializes this instance into a string.
   * @returns Encoded string.
   */
  public toString(): string {
    return this.toBuffer().toString('hex');
  }

  static fromString(str: string): Header {
    const buffer = Buffer.from(str.replace(/^0x/i, ''), 'hex');
    return Header.fromBuffer(buffer);
  }

  hash(): Fr {
    return pedersenHash(
      this.toFields().map(f => f.toBuffer()),
      GeneratorIndex.BLOCK_HASH,
    );
  }
}
