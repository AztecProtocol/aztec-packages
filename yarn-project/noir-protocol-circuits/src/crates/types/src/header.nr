use crate::{
    abis::{
        append_only_tree_snapshot::{
            AppendOnlyTreeSnapshot,
            APPEND_ONLY_TREE_SNAPSHOT_LENGTH,
        },
        global_variables::{
            GlobalVariables,
            GLOBAL_VARIABLES_LENGTH,
        },
    },
    constants::{
        GENERATOR_INDEX__BLOCK_HASH,
        HEADER_LENGTH,
        NUM_FIELDS_PER_SHA256,
        STATE_REFERENCE_LENGTH,
    },
    hash::pedersen_hash,
    state_reference::StateReference,
    traits::{
        Deserialize,
        Empty,
        Hash,
        Serialize,
    },
    utils::{
        arr_copy_slice,
    },
};

// docs:start:header
struct Header {
    last_archive: AppendOnlyTreeSnapshot,
    body_hash: [Field; NUM_FIELDS_PER_SHA256],
    state: StateReference,
    global_variables: GlobalVariables,
}
// docs:end:header

impl Serialize<HEADER_LENGTH> for Header {
    fn serialize(self) -> [Field; HEADER_LENGTH] {
        let mut fields: BoundedVec<Field, HEADER_LENGTH> = BoundedVec::new(0);

        fields.extend_from_array(self.last_archive.serialize());
        fields.extend_from_array(self.body_hash);
        fields.extend_from_array(self.state.serialize());
        fields.extend_from_array(self.global_variables.serialize());

        fields.storage
    }
}

impl Deserialize<HEADER_LENGTH> for Header {
    fn deserialize(serialized: [Field; HEADER_LENGTH]) -> Self {
        let mut offset = 0;

        let last_archive_fields = arr_copy_slice(serialized, [0; APPEND_ONLY_TREE_SNAPSHOT_LENGTH], offset);
        offset = offset + APPEND_ONLY_TREE_SNAPSHOT_LENGTH;

        let body_hash = arr_copy_slice(serialized, [0; NUM_FIELDS_PER_SHA256], offset);
        offset = offset + NUM_FIELDS_PER_SHA256;

        let state_fields = arr_copy_slice(serialized, [0; STATE_REFERENCE_LENGTH], offset);
        offset = offset + STATE_REFERENCE_LENGTH;

        let global_variables_fields = arr_copy_slice(serialized, [0; GLOBAL_VARIABLES_LENGTH], offset);

        Header {
            last_archive: AppendOnlyTreeSnapshot::deserialize(last_archive_fields),
            body_hash,
            state: StateReference::deserialize(state_fields),
            global_variables: GlobalVariables::deserialize(global_variables_fields),
        }
    }
}

impl Empty for Header {
    fn empty() -> Self {
        Self {
            last_archive: AppendOnlyTreeSnapshot::zero(),
            body_hash: [0; NUM_FIELDS_PER_SHA256],
            state: StateReference::empty(),
            global_variables: GlobalVariables::empty(),
        }
    }
}

impl Hash for Header {
    fn hash(self) -> Field {
        pedersen_hash(self.serialize(), GENERATOR_INDEX__BLOCK_HASH)
    }
}

#[test]
fn serialization_deserialization_smoke() {
    let header: Header = dep::std::unsafe::zeroed();
    let _serialized = header.serialize();
    let _deserialized = Header::deserialize(_serialized);
}

#[test]
fn hash_smoke() {
    let header: Header = dep::std::unsafe::zeroed();
    let _hashed = header.hash();
}
