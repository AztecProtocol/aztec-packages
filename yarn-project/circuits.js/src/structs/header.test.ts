import { HEADER_LENGTH } from '../constants.gen.js';
import { makeHeader } from '../tests/factories.js';
import { Header } from './header.js';

describe('Header', () => {
  let header: Header;

  beforeAll(() => {
    const randomInt = Math.floor(Math.random() * 1000);
    header = makeHeader(randomInt, undefined);
  });

  it('serializes to buffer and deserializes it back', () => {
    const buffer = header.toBuffer();
    const res = Header.fromBuffer(buffer);
    expect(res).toEqual(header);
  });

  it('serializes to field array and deserializes it back', () => {
    const fieldArray = header.toFields();
    const res = Header.fromFields(fieldArray);
    expect(res).toEqual(header);
  });

  it('computes hash', () => {
    const seed = 9870243;
    const header = makeHeader(seed, undefined);
    const hash = header.hash();
    expect(hash).toMatchSnapshot();
  });

  it('number of fields matches constant', () => {
    const fields = header.toFields();
    expect(fields.length).toBe(HEADER_LENGTH);
  });

  it('computes empty hash', () => {
    const header = Header.empty();
    const hash = header.hash();
    expect(hash).toMatchSnapshot();

    // Value used in empty_hash test in header.nr
    // console.log("hash", hash.toString());
  });
});
