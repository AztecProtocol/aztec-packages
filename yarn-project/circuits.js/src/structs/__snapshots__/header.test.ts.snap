// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header computes empty hash 1`] = `
Fr {
  "asBigInt": 20794359902049712309840803989537372971066838126069955186305683313093334375790n,
  "asBuffer": {
    "data": [
      45,
      249,
      48,
      204,
      123,
      159,
      199,
      99,
      232,
      42,
      222,
      114,
      247,
      196,
      97,
      136,
      52,
      105,
      43,
      42,
      61,
      9,
      54,
      175,
      248,
      215,
      187,
      251,
      39,
      245,
      157,
      110,
    ],
    "type": "Buffer",
  },
}
`;

exports[`Header computes hash 1`] = `
Fr {
  "asBigInt": 17965313985247589544372198735324565090920557482351371957268170919526618141863n,
  "asBuffer": {
    "data": [
      39,
      184,
      1,
      227,
      84,
      195,
      178,
      90,
      6,
      248,
      237,
      161,
      177,
      253,
      166,
      87,
      161,
      67,
      151,
      185,
      244,
      147,
      251,
      142,
      40,
      36,
      177,
      92,
      201,
      16,
      148,
      167,
    ],
    "type": "Buffer",
  },
}
`;
